extcomp-template/CMakeLists.txt

```diff
@@ -0,0 +1,37 @@
+# Modify this list as necessary to include all your source files.
+set(extmod_srcs
+  "mymod.c"
+)
+
+# If necessary, add items to the PRIV_REQUIRES list below
+idf_component_register(
+  SRCS ${extmod_srcs}
+  INCLUDE_DIRS "." "${CMAKE_CURRENT_BINARY_DIR}"
+  PRIV_REQUIRES
+    "base_nodemcu"
+    "lua"
+    "platform"
+)
+
+# The remainder is boiler-plate glue to get the linker to actually include
+# the modules enabled in Kconfig. No user-serviceable parts inside.
+
+# Match up all the extmod source files with their corresponding Kconfig
+# option in the form NODEMCU_CMODULE_<modname> and if enabled, add a
+# "-u <modname>_module_selected1" option to force the linker to include
+# the module. See components/core/include/module.h for further details on
+# how this works.
+set(extmods_enabled)
+foreach(extmod_src ${extmod_srcs})
+  string(REPLACE ".c" "" module_name ${extmod_src})
+  string(TOUPPER ${module_name} module_ucase)
+  set(mod_opt "CONFIG_NODEMCU_CMODULE_${module_ucase}")
+  if (${${mod_opt}})
+    list(APPEND extmods_enabled ${module_ucase})
+  endif()
+endforeach()
+message("Including the following modules: ${extmods_enabled}")
+
+foreach(mod ${extmods_enabled})
+  target_link_libraries(${COMPONENT_LIB} "-u ${mod}_module_selected1")
+endforeach()
```

Comment:
Maybe? I have to admit that I don't know how scoping works across different components in the IDF and whether it's possible to expose a function in one for use in another one. I'm pretty handy with a regular Makefile, but CMake it an entirely different kettle of fish that I'm not comfortable with yet.

Looking at the way I've implemented this, I'm also wondering whether it's possible to instead iterate over the sdkconfig variables that start with `CONFIG_NODEMCU_CMODULE_` instead of source files...

Reply:
Okay, that definitely looks cleaner! I'd say the function should probably go into a `project_include.cmake` file in the `components/modules/` directory though, as it relates to module building.

I wonder if it would work to do the filtering of the `CONFIGS_LIST` once in the regular module directory, and then (if necessary) have external modules just depend on "modules"; whether that would get the `-u XYZ_module_selected1` onto the linker command line in the right place...